### ISSUE TITLE
Recursion limit during encode

### ISSUE THREAD

**Issue by crepererum** [tokio-rs/prost#736]
The decode routines have a recursion limit (currently hard-coded to 100 if the `no-recursion-limit` is NOT set), however the encode routines don't. This has two drawbacks:

1. **Stack Overflow:** It is easy to overflow the stack during encoding (however this is less severe than during decoding, because the data at hand is likely "more trusted").
2. **Roundtrip Safety:** Within the same software stack, it is possible to encode a message that cannot be decoded anymore. This is somewhat surprising.

It would be nice if encoding a message would use the same recursion limit checker.

Context: https://github.com/apache/arrow-datafusion/issues/3968

**Comment by LucioFranco** [comment]
Yeah, I think it makes a lot of sense! I don't have much time now to get to this but I would accept a PR for it.

**Comment by liukun4515** [comment]
but now, if we can encode the data, but want to decode the data right, what should we do?
enable the feature `no-recursion-limit` @crepererum
